Refactor Daily Navigation using Keybindings.

sshanzel opened the pull request:
Dear Team,

As the design had major changes, the DOM structure was modified as well. This PR is in relation to the update, to ensure the feature would function as expected for the major release. 

Hi @idoshamun, I've used the `offsetTop` property to identify the mapping of the `posts` in the `feed`, as I realized there's a pattern in the current design.

I feel that this PR is ready for review though I am still writing multiple test cases for this at the moment to ensure maintainability and integrity of the feat.

Please let me know if you have any feedback or concerns about the current refactoring.

Best Regards,
Lee

idoshamun commented on packages/extension/src/common/keyNavigationService.js:
```diff
@@ -17,75 +17,56 @@ export const validKeys = {
+  const ref = daFeedRef();
 
-  const item = getCurrentPost(posts, current);
+  if (keyCode === validKeys.esc) return store.dispatch('feed/backToMainFeed');
```
I think here we should also reset the "show bookmarks" state. I often click `esc` while browsing my bookmarks and it just refreshes the feed instead of going back to the main feed. What do you think?

sshanzel replied:
Will definitely add this on the list! 💯

idoshamun commented on packages/extension/src/common/keyNavigationService.js:
```diff
@@ -17,75 +17,56 @@ export const validKeys = {
-
-function navigate(keyCode, posts, search, backToMainFeed, { current, insaneMode }) {
-  if (keyCode === validKeys.esc) return backToMainFeed();
+  const { daFeedRef, hoveredPost } = store.state.feed;
```
I don't like saving references in the store. Maybe it's better to initialize the key navigation service with callback functions and once something happens instead of using the refs, the service will call these callbacks.
I find current implementation too coupled to the components and their functions. By using callbacks it will be agnostic.

sshanzel replied:
I felt too that it is having dependencies on so many places especially the reference of a whole component. Will definitely work on your suggestion in making it agnostic.

idoshamun commented on packages/components/src/common/domHelper.js:
```diff
@@ -1,33 +1,89 @@
 
-export function getAbovePost(el) {
-  if (el.previousElementSibling === null) return el;
+export function getElementIndexFromSiblings(targetElement) {
```
Can we try to think about an implementation that doesn't require looping through all previous elements to calculate the index? I fear that in extreme cases this loop can slow things down.
Maybe we can save the index of the selected post instead of saving the reference to the element. I'm open to any other ideas. Let me know.

sshanzel replied:
Yes I think we can do that. I'll probably create a map when they are rendered so we won't have to recompute everytime, identifying of the index will be part of the render itself that will have an O(1) when finding and saving its index.

idoshamun replied:
Yes, I'm looking for an O(1) solution. That will be perfect. I'm here if you want to discuss it over.

sshanzel replied:
Will definitely ping you up once I hit a brick wall 😄 Thank you Ido!

sshanzel replied:
Hi @idoshamun, what do you think if I expand the `hoveredPost` reference in making it an array having the current implementation value at index `0` and the actual index of the element at index `1` so it won't have to recompute every time?

idoshamun replied:
That sounds interesting! We can try it out

sshanzel replied:
Will create a PR soon!

sshanzel commented:
Hi @idoshamun, thank you for the kind words, I really appreciate it!

I've made the necessary changes for the feature to be available again.

- I've extended the `backToMainFeed` to allow going back if current feed is at the Bookmarks page.

- I've removed the dependency of the `keyNavigationService.js` from the store complete and removed the `DaFeed` component reference from the store as well.

- I've injected a function to the `textarea` for the comment so the `DailyNavigation` will be disabled. The `Home` route is watching its value and make the necessary decision whether to turn on the feature or not.

- With regards to the issue at `getElementIndexFromSiblings` looping through the DOM, Unfortunately, my trials of converting a map for it to reference failed. I tried creating a map having `index` as the key and `post`  component reference as the value and vice versa (this is generated whenever `posts` at the store changes). There's an issue arising due to the indexes are not matching with the iteration from `template` when rendering `posts` components (I don't know yet if it is possible with Vue, to have a function that will return an HTML `template` so I can synchronously map the indexes). I stashed the changes I've made so I can easily pull it up.

Thank you very much for the patience brother.

If you have any concerns or issues with the current changes, please let me know so I can take further action.

Sorry for the delay!

Best Regards,
Lee

sshanzel commented:
I have thought of something, it can be expensive when the `posts` changes, but I'll check it out.

idoshamun commented on packages/extension/src/common/keyNavigationService.js:
```diff
@@ -17,88 +16,48 @@ export const validKeys = {
+  const nextPostAfterEmptyAdElement = getNextPost(foundElement, insaneMode);
 
-function navigateDaily({ keyCode, target }) {
-  if (target instanceof HTMLInputElement || target instanceof HTMLTextAreaElement) {
-    return;
-  }
```
Instead of manually turning on/off the keybindings. This code snippet can detect automatically if the keypress is targeted to an input element.
I think it's a better solution as it is future-proof.

sshanzel replied:
Ohhhh, that's amazing, I never knew this. Thank you for the info!

idoshamun replied:
You bet! I've implemented it a while ago 🤩

sshanzel replied:
Yeah, I saw the changes. Sincere apologies for omitting! I didn't realize it was for that.

idoshamun commented:
> I have thought of something, it can be expensive when the `posts` changes, but I'll check it out. It will be an O(n²) when the `posts` changes. That will be the trade-off for the O(1) in getting the `elementSiblingsIndex`. What do you think?

So let's keep it like before. This might be better.

idoshamun commented:
Please also make sure to resolve the conflicts. If you need help let me know 💪

sshanzel commented:
Hi @idoshamun, thank you for the kind comments, I think this is now ready for review. Thank you!

idoshamun commented:
Woohoo! Review it as soon as I can

sshanzel commented:
Thank you so much Ido for understanding the time has taken me! Best of luck!

idoshamun commented on packages/extension/src/common/keyNavigationService.js:
```diff
@@ -17,88 +16,48 @@ export const validKeys = {
-  window.addEventListener('keydown', navigateDaily);
+export function bindEvent(event, callback) {
+  window.addEventListener(event, callback);
 }
 
-export function disableKeyBindings() {
-  window.removeEventListener('keydown', navigateDaily);
+export function unbindEvent(event, callback) {
+  window.removeEventListener(event, callback);
 }
```
I think these functions are no longer needed.
Other than that it looks great!

sshanzel replied:
I didn't realize I wrapped an existing function with the same thing as it does. Lol! 😆 Too much abstraction my friend 🤣

idoshamun reviewed (changes requested):
Last change and we can merge!
Thank you 🙏

sshanzel commented:
Hi @idoshamun, I have now omitted the event binders. Thank you for the awesome reviews! 🍻 🎉

idoshamun reviewed (approved):
We did it! 🚀
Hopefully, I will launch it with the next version

sshanzel commented:
Yayy! We did it! 🎉

idoshamun commented:
That's great! Thanks :)

On Sat, Aug 29, 2020 at 6:20 PM Hansel Solevilla <notifications@github.com>
wrote:

> *@sshanzel* commented on this pull request.
> ------------------------------
>
> In packages/components/src/common/domHelper.js
> <https://github.com/dailydotdev/daily-apps/pull/148#discussion_r479659532>
> :
>
> >  }
>
> -export function getAbovePost(el) {
> -  if (el.previousElementSibling === null) return el;
> +export function getElementIndexFromSiblings(targetElement) {
>
> Will create a PR soon!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dailydotdev/daily-apps/pull/148#discussion_r479659532>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPGUHIQSZTQAJRPDYZM263SDEMELANCNFSM4P7G7KGA>
> .
>
